chore: Update prisma to Prisma 3

emrysal opened the pull request:
For background info and discussion on referential actions check #970 by @KATT 

Please read https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions

### Todos:

- [x] Review 
- [ ] Definitely will need tweaks 

* Tweak 1: Turned Payment booking foreign key to SetNull when the Booking goes *poof*
* Tweak 2: Turned Booking with event type foreign key to SetNull when the EventType goes *poof*

- [ ] Review again
- [ ] Create new migration file (A preview can be seen here)
- [ ] Merge
- [ ] Make sure that migration is run at the time of merging - @baileypumfleet / @emrysal

emrysal committed:
chore: Update prisma to Prisma 3

emrysal committed:
Implements native Prisma seeding features with our seed.ts

emrysal commented:
@KATT This also makes our seed.ts script compatible with Prisma Seeding @ https://www.prisma.io/docs/guides/database/seed-database

emrysal committed:
Add prisma directory to lint-staged

KATT commented on package.json:
```diff
@@ -22,6 +21,9 @@
     "prepare": "husky install"
   },
+  "prisma": {
+    "seed": "ts-node --compiler-options {\"module\":\"CommonJS\"} ./prisma/seed.ts"
```
This is is already in the `ts-node` script above. Should be enough with 

```suggestion
    "seed": "ts-node ./prisma/seed.ts"
```

or possibly


```suggestion
    "seed": "yarn ts-node ./prisma/seed.ts"
```

KATT reviewed (commented):
Questions

- Does the CI build break if we were to update to Prisma `3.1.1` - if not, it would be good if it does so no one accidentally upgrades it
- The seed script doesn't run on normal prisma migrations, right? We wouldn't want the seed to be in prod, ever

emrysal committed:
Update package.json

Use yarn ts-node instead of ts-node which already has the correct compiler options

Co-authored-by: Alex Johansson <alexander@n1s.se>

emrysal committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

emrysal committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

KATT committed:
Merge remote-tracking branch 'origin/main' into tweak/update-prisma-to-3.0.2

# Conflicts:
#	package.json

KATT commented on .github/workflows/e2e.yml:
```diff
@@ -54,8 +54,7 @@ jobs:
       - run: yarn test
-      - run: yarn prisma migrate deploy
-      - run: yarn db-seed
+      - run: yarn prisma migrate reset --force
```
I think this should be reverted?

KATT reviewed (changes requested):
Missing migrations that prisma makes as part of the upgrade - try running `yarn db-migrate` locally or `yarn db-nuke && yarn dx` and you'll be prompted

This is what prisma does in that migration -- not thought if this is a good or bad thing yet. Personally, I like cascade deletes on non-critical data as it's easier to work with, but seems like we need to update [referential actions](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions) as part of this upgrade.

<details>

```sql
-- DropForeignKey
ALTER TABLE "EventTypeCustomInput" DROP CONSTRAINT "EventTypeCustomInput_eventTypeId_fkey";

-- DropForeignKey
ALTER TABLE "Membership" DROP CONSTRAINT "Membership_teamId_fkey";

-- DropForeignKey
ALTER TABLE "Membership" DROP CONSTRAINT "Membership_userId_fkey";

-- DropForeignKey
ALTER TABLE "Payment" DROP CONSTRAINT "Payment_bookingId_fkey";

-- DropForeignKey
ALTER TABLE "SelectedCalendar" DROP CONSTRAINT "SelectedCalendar_userId_fkey";

-- AddForeignKey
ALTER TABLE "Membership" ADD CONSTRAINT "Membership_teamId_fkey" FOREIGN KEY ("teamId") REFERENCES "Team"("id") ON DELETE RESTRICT ON UPDATE CASCADE;

-- AddForeignKey
ALTER TABLE "Membership" ADD CONSTRAINT "Membership_userId_fkey" FOREIGN KEY ("userId") REFERENCES "users"("id") ON DELETE RESTRICT ON UPDATE CASCADE;

-- AddForeignKey
ALTER TABLE "SelectedCalendar" ADD CONSTRAINT "SelectedCalendar_userId_fkey" FOREIGN KEY ("userId") REFERENCES "users"("id") ON DELETE RESTRICT ON UPDATE CASCADE;

-- AddForeignKey
ALTER TABLE "EventTypeCustomInput" ADD CONSTRAINT "EventTypeCustomInput_eventTypeId_fkey" FOREIGN KEY ("eventTypeId") REFERENCES "EventType"("id") ON DELETE RESTRICT ON UPDATE CASCADE;

-- AddForeignKey
ALTER TABLE "Payment" ADD CONSTRAINT "Payment_bookingId_fkey" FOREIGN KEY ("bookingId") REFERENCES "Booking"("id") ON DELETE RESTRICT ON UPDATE CASCADE;

-- RenameIndex
ALTER INDEX "Booking.uid_unique" RENAME TO "Booking_uid_key";

-- RenameIndex
ALTER INDEX "EventType.userId_slug_unique" RENAME TO "EventType_userId_slug_key";

-- RenameIndex
ALTER INDEX "Payment.externalId_unique" RENAME TO "Payment_externalId_key";

-- RenameIndex
ALTER INDEX "Payment.uid_unique" RENAME TO "Payment_uid_key";

-- RenameIndex
ALTER INDEX "Team.slug_unique" RENAME TO "Team_slug_key";

-- RenameIndex
ALTER INDEX "VerificationRequest.identifier_token_unique" RENAME TO "VerificationRequest_identifier_token_key";

-- RenameIndex
ALTER INDEX "VerificationRequest.token_unique" RENAME TO "VerificationRequest_token_key";

-- RenameIndex
ALTER INDEX "Webhook.id_unique" RENAME TO "Webhook_id_key";

-- RenameIndex
ALTER INDEX "users.email_unique" RENAME TO "users_email_key";

-- RenameIndex
ALTER INDEX "users.username_unique" RENAME TO "users_username_key";

```
</details>

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

kodiakhq[bot] committed:
Merge branch 'main' into tweak/update-prisma-to-3.0.2

KATT commented:
Converted this to draft so it's not accidentally merged

KATT committed:
Add referential actions (prisma 3 upgrade) (#970)

* set referential actions

these needs to be scrutinized and thought about in a lot more detail

* tweak

* remove referential actions on array

* migration preview

* Revert "migration preview"

This reverts commit 1631ca283c4860dc8a18b25ae01447f96f80897e.

* fix seed

emrysal committed:
Prisma 3.0.2 -> 3.3.0

emrysal committed:
Merged main into tweak/update-prisma-to-3.0.2

emrysal committed:
Set Payment, Booking to SetNull on foreign key delete

mohammedzamakhan commented on package.json:
```diff
@@ -136,14 +139,14 @@
     "postcss": "^8.3.11",
     "prettier": "^2.3.2",
-    "prisma": "^2.30.2",
+    "prisma": "^3.3.0",
```
3.5.0 just got released!
